kernel/src/table_changes/mod.rs

```diff
@@ -0,0 +1,191 @@
+//! In-memory representation of a change data feed table.
+
+use std::{collections::HashMap, sync::Arc};
+
+use crate::{
+    actions::{Metadata, Protocol},
+    features::ColumnMappingMode,
+    log_segment::{LogSegment, LogSegmentBuilder},
+    path::AsUrl,
+    scan::{get_state_info, state::DvInfo, ColumnType},
+    schema::{Schema, SchemaRef, StructType},
+    snapshot::Snapshot,
+    DeltaResult, Engine, EngineData, Error, ExpressionRef, Version,
+};
+use url::Url;
+
+pub type TableChangesScanData = (Box<dyn EngineData>, Vec<bool>, Arc<HashMap<String, DvInfo>>);
+
+static CDF_ENABLE_FLAG: &str = "delta.enableChangeDataFeed";
+
+#[derive(Debug)]
+pub struct TableChanges {
+    pub snapshot: Snapshot,
+    #[allow(unused)]
+    pub(crate) log_segment: LogSegment,
+    pub schema: Schema,
+    pub version: Version,
+    pub metadata: Metadata,
+    pub protocol: Protocol,
+    pub(crate) column_mapping_mode: ColumnMappingMode,
+    pub table_root: Url,
+}
+
+impl TableChanges {
+    pub fn try_new(
+        table_root: Url,
+        engine: &dyn Engine,
+        start_version: Version,
+        end_version: Option<Version>,
+    ) -> DeltaResult<Self> {
+        let start_snapshot =
+            Snapshot::try_new(table_root.as_url().clone(), engine, Some(start_version))?;
+        let end_snapshot = Snapshot::try_new(table_root.as_url().clone(), engine, end_version)?;
+        println!("Start snapshot: {:?}", start_snapshot);
+        println!("End snapshot: {:?}", end_snapshot);
+
+        let start_flag = start_snapshot.metadata().configuration.get(CDF_ENABLE_FLAG);
+        let end_flag = end_snapshot.metadata().configuration.get(CDF_ENABLE_FLAG);
+
+        // Verify CDF is enabled at the beginning and end of the interval
```

Comment:
I believe (correct me if I'm wrong) but CDF needs to remain enabled for the ENTIRE span of the versions they are requesting. I do not think it's enough to just validate the start and end.

Reply:
Correct. But that should be done later when we're replaying the log. At this stage, we do a best effort.